### PR TITLE
Add C++11 range utilities

### PR DESCRIPTION
**Pull request by K-ballo**

### CODE REVIEW

File: hpx/lcos/when_all.hpp [hunk 1]
```diff
@@ -356,46 +357,43 @@ namespace hpx { namespace lcos
...
     }
 
-    template <typename Range>
-    typename std::enable_if<traits::is_future_range<Range>::value,
-        lcos::future<typename std::decay<Range>::type> >::type
-    when_all(Range& values)
-    {
-        Range values_ = traits::acquire_future<Range>()(values);
-        return lcos::when_all(std::move(values_));
-    }
```
Comment (sithhell) [inline hpx/lcos/when_all.hpp]: What's the rationale to drop this overload?

File: hpx/lcos/when_all.hpp [hunk 2]
```diff
@@ -356,46 +357,43 @@ namespace hpx { namespace lcos
...
+        result_type lazy_values_ = traits::acquire_future<result_type>()(lazy_values);
+
+        typedef detail::when_all_frame<
+                util::tuple<result_type>
+            > frame_type;
         typedef typename frame_type::init_no_addref init_no_addref;
 
         boost::intrusive_ptr<frame_type> p(new frame_type(
-            util::forward_as_tuple(std::move(values)), init_no_addref()),
+            util::forward_as_tuple(std::move(lazy_values_)), init_no_addref()),
```
Comment (sithhell) [inline hpx/lcos/when_all.hpp]: Shouldn't we rather `std::forward` the lazy values?

**Comment by K-ballo** [comment]
> Did you measure any compile time improvements with this patch?

No, no measurement was performed. Compile time improvements is not the rationale of this PR, though it'd be a colorful datapoint to have if someone's interesting in doing the work.

File: hpx/lcos/when_all.hpp (hunk 1 above)
Comment (K-ballo) [inline hpx/lcos/when_all.hpp]: This overload becomes redundant, given the changes to the overload above.

File: hpx/lcos/when_all.hpp (hunk 2 above)
Comment (K-ballo) [inline hpx/lcos/when_all.hpp]: No. That might have made some sense (not really) back when `std::vector` was the only thing accepted as a range, but now the function will accept non-owning ranges as well, and needs to acquire the shared states explicitly so as to guarantee they stay alive long enough.

**Comment by sithhell** [comment]
Yes, I can see that a reduction of compile times wasn't the point here. A difference in compile times would be interesting nonetheless.

File: hpx/lcos/when_all.hpp (hunk 2 above)
Comment (sithhell) [inline hpx/lcos/when_all.hpp]: Then I am a little confused. The function takes a forwarding reference, yet you move it unconditionally. How does it work? Wouldn't `when_all_frame` capture it correctly even if it was forwarded?

Comment (K-ballo) [inline hpx/lcos/when_all.hpp]: We don't move from the argument `lazy_values`, but from a local `lazy_values_` initialized by acquiring all the futures in the range given as argument. This is rather subtle, but a common pattern within the `when`s.

That said, it looks like the implementation of `acquire_future` might be insufficient to properly support non-owning ranges. I'll have to look into that.

Comment (sithhell) [inline hpx/lcos/when_all.hpp]: Ahh, thanks for the clarification. I missed the underscore...

**Pull request merged**